network/common/src/main/java/org/apache/spark/network/buffer/LargeByteBuffer.java

```diff
@@ -0,0 +1,135 @@
+  public byte get();
+
+  /**
+   * Bulk copy data from this buffer into the given array.  First checks there is sufficient
+   * data in this buffer; if not, throws a {@link java.nio.BufferUnderflowException}.
+   *
+   * @param dst
+   * @param offset
+   * @param length
+   */
+  public void get(byte[] dst, int offset, int length);
+
+  public LargeByteBuffer rewind();
+
+  /**
+   * Return a deep copy of this buffer.
+   * The returned buffer will have position == 0.  The position
+   * of this buffer will not change as a result of copying.
+   *
+   * @return a new buffer with a full copy of this buffer's data
+   */
+  public LargeByteBuffer deepCopy();
+
+  /**
+   * Advance the position in this buffer by up to <code>n</code> bytes.  <code>n</code> may be
+   * positive or negative.  It will move the full <code>n</code> unless that moves
+   * it past the end (or beginning) of the buffer, in which case it will move to the end
+   * (or beginning).
+   *
+   * @return the number of bytes moved forward (can be negative if <code>n</code> is negative)
+   */
+  public long skip(long n);
+
+  public long position();
+
+  /**
+   * Creates a new byte buffer that shares this buffer's content.
+   * <p/>
+   * The content of the new buffer will be that of this buffer.  Changes
+   * to this buffer's content will be visible in the new buffer, and vice
+   * versa; the two buffers' positions will be independent.
+   * <p/>
+   * The new buffer's position will be identical to those of this buffer
+   */
+  public LargeByteBuffer duplicate();
+
+  public long remaining();
+
+  /**
+   * Total number of bytes in this buffer
+   */
+  public long size();
+
+  /**
+   * Writes the data from the current <code>position()</code> to the end of this buffer
+   * to the given channel.  The <code>position()</code> will be moved to the end of
+   * the buffer after this.
+   * <p/>
+   * Note that this method will continually attempt to push data to the given channel.  If the
+   * channel cannot accept more data, this will continuously retry until the channel accepts
+   * the data.
+   *
+   * @param channel
+   * @return the number of bytes written to the channel
+   * @throws IOException
+   */
+  public long writeTo(WritableByteChannel channel) throws IOException;
+
+  /**
+   * Get the entire contents of this as one ByteBuffer, if possible.  The returned ByteBuffer
+   * will always have the position set to 0, and the limit set to the end of the data.  Each
+   * call will return a new ByteBuffer, but will not require copying the data (eg., it will
+   * use ByteBuffer#duplicate()).  The returned byte buffer will share data with this buffer.  The
+   * returned buffers will never be larger than
+   * {@link org.apache.spark.network.buffer.LargeByteBufferHelper#MAX_CHUNK_SIZE}
+   *
+   * @throws BufferTooLargeException if this buffer is too large to fit in one {@link ByteBuffer}
+   */
+  public ByteBuffer asByteBuffer() throws BufferTooLargeException;
```

Comment:
Eventually we may no longer use `ByteBuffer` directly, but initially some parts of the code will still use it.  In particular, we'll start by only supporting caching for large blocks, we won't support replication or shuffling (because its more complicated w/ the network transfer).  So those parts of the code will still need this.  See this discussion on https://issues.apache.org/jira/browse/SPARK-1391